<a href="https://colab.research.google.com/github/HwangHanJae/ml-definitive-guide-pratice/blob/main/pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

read_csv()의 특징 :
- 필드 구분 문자가 콤마(',')로 구분
- (파일명, sep ='\t')으로 탭으로 구분된 데이터 로드 가능

read_table()의 특징 : 
- 필드 구분 문자가 탭('\t')으로 구분

read_fwf()의 특징 : 
- Fixed Width 고정 길이 기반의 칼럼 포맷

In [112]:
import pandas as pd

#캐글 - 타이타닉 탑승자 데이터
#read_csv()로 읽어오기, (filepath_of_buffer, sep = ','...)
#filepath_or_buffer 자리에는 파일 경로

file_path = "/content/drive/MyDrive/Colab Notebooks/DataScience/Data/titanic/train.csv"
titanic_df = pd.read_csv(file_path)
print("titalnic_df 변수 type : ", type(titanic_df))

#head()는 defalut는 5, 인자값으로 n을 지정하면 n만큼 n개의 로우를 반환
titanic_df.head()
#index(0,1,3,4...)객체는 sql의 pk와 유사하게 고유의 레코드를 식별하는 역할

titalnic_df 변수 type :  <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
#shape : 생성된 DataFrame 객체의 크기를 확인

print("titanic_df 의 크기 : ",titanic_df.shape)
print("생성된 titanic_df는 {}개의 행(로우)과 {}개의 열(칼럼)을 가지고 있습니다.".format(titanic_df.shape[0], titanic_df.shape[1]))

titanic_df 의 크기 :  (891, 12)
생성된 titanic_df는 891개의 행(로우)과 12개의 열(칼럼)을 가지고 있습니다.


In [ ]:
#info() : 총 데이터 건수와 데이터타입, Null 건수를 확인
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info() 해석
1. RangeIndex는 DataFrame index의 범위를 나타냅니다. 
- 전체 로우의 수를 알 수 있습니다. 데이터는 891개 로우 입니다. 그리고 컬럼수는 12개입니다.

2. Dtype은 칼럼별 타입을 나타냅니다.
- Pclass 칼럼의 데이터 타입은 int64형 입니다.
- Name 칼럼의 데이터 타입은 object입니다. 이는 문자열타입으로 생각해도 무방합니다.

3. Non-Null Count를 보면 몇개의 데이터가 non-null인지 나타냅니다.
- Age 칼럼의 714 non-null의 의미는 Age 칼럼 891개 데이터 중 714개가 Null이 아니며 177개는 Null이라는 의미입니다.

4. dtype : 은 전체 12개의 칼럼들의 타입을 요약한 것 입니다.
- 2개 칼럼이 float64, 5개의 칼럼이 int64, 5개 칼럼이 object 타입입니다.


In [ ]:
#describe() 오직 숫자형(int, float 등)칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외합니다.
titanic_df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

In [ ]:

value_counts = titanic_df["Pclass"].value_counts()

print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


In [ ]:
#DataFrame의 []연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환
titanic_pclass = titanic_df["Pclass"]

print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [ ]:
#Series 타입에도 head() 메서드를 사용할 수 있음
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [ ]:
#넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환
import numpy as np

col_name1 =['co1']

#리스트를 이용해 DataFrame 생성
list1 = [1,2,3]
df_list1 = pd.DataFrame(list1, columns=col_name1)
print("1차원 리스트로 만든 DataFrame : \n ", df_list1)

#넘파이 ndarray를 이용해 DataFrame 생성.
array1 = np.array(list1)
df_array1 = pd.DataFrame(array1, columns=col_name1)
print("1차원 ndarray로 만든 DataFrame : \n", df_array1)

1차원 리스트로 만든 DataFrame : 
     co1
0    1
1    2
2    3
1차원 ndarray로 만든 DataFrame : 
    co1
0    1
1    2
2    3


In [ ]:
#2차원 형태의 데이터를 기반으로 DataFrame생성

#3개의 칼럼명이 필요함
col_name2 = ['col1', 'col2','col3']

#2행 x 3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환

list2 = [[1,2,3],
         [11,22,33]]

array2 = np.array(list2)
#array2 의 크기 확인
print("array2 의 크기 : ", array2.shape)
print("array2는 {}행 x {}열의 형태이다. \n".format(array2.shape[0], array2.shape[1]))

#리스트를 이용하여 DataFrame 만들기
df_list2 = pd.DataFrame(data=list2, columns=col_name2)
print("2차원 리스트로 만든 DataFrame : \n", df_list2)

#ndarray를 이용하여 DataFrame 만들기
df_array2 = pd.DataFrame(data=array2, columns=col_name2)
print("2차원 ndarray로 만든 DataFrame : \n", df_array2)

array2 의 크기 :  (2, 3)
array2는 2행 x 3열의 형태이다. 

2차원 리스트로 만든 DataFrame : 
    col1  col2  col3
0     1     2     3
1    11    22    33
2차원 ndarray로 만든 DataFrame : 
    col1  col2  col3
0     1     2     3
1    11    22    33


In [ ]:
#딕셔너리를 DataFrame으로 변환
#딕셔너리의 key는 칼럼, value는 키에 해당하는 칼럼 데이터로 변환

#key는 문자열, value는 리스트형(또는 ndarray)칼럼 데이터로 매핑
dict1 = {'col1' : [1,2,3], "col2" : [11,22,33], "col3" : [111,222,333]}
df_dict = pd.DataFrame(dict1)

print("딕셔너리로 만든 DataFrame : \n", df_dict)

딕셔너리로 만든 DataFrame : 
    col1  col2  col3
0     1    11   111
1     2    22   222
2     3    33   333


In [ ]:
#많은 머신러닝 패키지는 기본 데이터 형으로 ndarray를 사용함
#따라서 DataFrame을 ndarray로 변환할 수 있어야 함
#DataFrame 객체의 values를 이용할 수 있음

#DataFrame을 ndarray로 변환
array3 = df_dict.values
print("df_dict.values 타입 : ", type(array3))
print("df_dict.values의 크기 : ", array3.shape)
#array3을 출력
print(array3)

df_dict.values 타입 :  <class 'numpy.ndarray'>
df_dict.values의 크기 :  (3, 3)
[[  1  11 111]
 [  2  22 222]
 [  3  33 333]]


In [ ]:
df = df_dict.copy()

#DataFrame을 리스트와 딕셔너리로 변환

#DataFrame -> list
#세부 : DataFrame -> ndarray -> list
list3 = df.values.tolist()
print("df.values.tolist()의 타입 : ", type(list3))
print(list3)

#DataFrame -> dict
#to_dict()의 인자값은 'list'로 지정하면 딕셔너리의 value가 list로 변환됨
dict3 = df.to_dict('list')
print("df.values.to_dict()의 타입 : ", type(dict3))
print(dict3)

df.values.tolist()의 타입 :  <class 'list'>
[[1, 11, 111], [2, 22, 222], [3, 33, 333]]
df.values.to_dict()의 타입 :  <class 'dict'>
{'col1': [1, 2, 3], 'col2': [11, 22, 33], 'col3': [111, 222, 333]}


In [ ]:
#DataFrame의 칼럼 데이터 세트 생성과 수정
titanic_df["Age_0"] = 0

titanic_df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Age_0
0            1         0       3  ...   NaN        S      0
1            2         1       1  ...   C85        C      0
2            3         1       3  ...   NaN        S      0
3            4         1       1  ...  C123        S      0
4            5         0       3  ...   NaN        S      0

[5 rows x 13 columns]

In [ ]:
#기존 칼럼을 이용하여 새로운 칼럼 생성

#Age를 이용
titanic_df["Age_by_10"] = titanic_df["Age"] * 10

#SibSp와 Parch를 이용
titanic_df["Family_No"] = titanic_df["SibSp"] + titanic_df["Parch"] + 1

#결과 확인
titanic_df.head()

PassengerId  Survived  Pclass  ... Age_0 Age_by_10  Family_No
0            1         0       3  ...     0     220.0          2
1            2         1       1  ...     0     380.0          2
2            3         1       3  ...     0     260.0          1
3            4         1       1  ...     0     350.0          2
4            5         0       3  ...     0     350.0          1

[5 rows x 15 columns]

In [ ]:
#기존 칼럼값 변경
#Age_by_10 변경
titanic_df["Age_by_10"] = titanic_df["Age_by_10"] + 100
titanic_df.head()

PassengerId  Survived  Pclass  ... Age_0 Age_by_10  Family_No
0            1         0       3  ...     0     320.0          2
1            2         1       1  ...     0     480.0          2
2            3         1       3  ...     0     360.0          1
3            4         1       1  ...     0     450.0          2
4            5         0       3  ...     0     450.0          1

[5 rows x 15 columns]

drop() 메서드의 원형 :
- DataFrame.drop(**labels**=None, **axis=0**, index=None, columns=None, level=None, **inplace=False**, errors ='raise')
  - labels : 원하는 칼럼명
  - axis : 0은 로우 방향축, 1은 칼럼 방향 축
  - inplace : True면 원본 DataFrame 삭제

In [ ]:
#DataFrame 데이터 삭제

#drop() 메서드를 이용합니다., inplace = False

titanic_drop_df = titanic_df.drop("Age_0", axis=1)
titanic_drop_df.head()

PassengerId  Survived  Pclass  ... Embarked Age_by_10  Family_No
0            1         0       3  ...        S     320.0          2
1            2         1       1  ...        C     480.0          2
2            3         1       3  ...        S     360.0          1
3            4         1       1  ...        S     450.0          2
4            5         0       3  ...        S     450.0          1

[5 rows x 14 columns]

In [ ]:
drop_result = titanic_df.drop(['Age_0', "Age_by_10", "Family_No"], axis=1, inplace=True)
print("inplce = True 로 drop 후 반환된 값 : ", drop_result)
titanic_df.head()

inplce = True 로 drop 후 반환된 값 :  None


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
#axis=0 을 사용하여 index 0,1,2(맨 앞 3개 데이터) 로우를 삭제
print("####before drop axis=0####")
print(titanic_df.head())
print("####after drop axis=0####")
print(titanic_df.drop([0,1,2], axis=0).head())

####before drop axis=0####
   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]
####after drop axis=0####
   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S
5            6         0       3  ...   8.4583   NaN         Q
6            7         0       1  ...  51.8625   E46         S
7            8         0       3  ...  21.0750   NaN         S

[5 rows x 12 columns]


In [ ]:
#인덱스 객체
#DataFrame, Seires 데이터에서 Index 객체만 추출
indexes = titanic_df.index
print("Index 객체 : ",indexes)
#index 객체를 실제 값 array로 변환
print("indexes.values 타입  : ", type(indexes.values))
print("Index 객체 array값 : \n", indexes.values)

Index 객체 :  RangeIndex(start=0, stop=891, step=1)
indexes.values 타입  :  <class 'numpy.ndarray'>
Index 객체 array값 : 
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217

In [ ]:
#Index객체는 단일 값 반환 및 슬라이싱 가능
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


In [ ]:
#Index 객체는 함부로 변경할 수 없음
indexes[0] = 5
#Index does not support mutable operations

TypeError: ignored

In [ ]:
#Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때
#Index는 연산에서 제외됩니다.

series_fair = titanic_df["Fare"]
print("Fair Series max 값 : ", series_fair.max())
print("Fair Series min 값 : ,", series_fair.min())
print("sum() Fair Series : ", series_fair.sum())
print("Fair Series + 3 :\n", (series_fair + 3).head(3))

Fair Series max 값 :  512.3292
Fair Series min 값 : , 0.0
sum() Fair Series :  28693.9493
Fair Series + 3 :
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [ ]:
#reset_index() : 새롭게 인덱스를 연속 숫자 형으로 할당
#기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가
titanic_reset_df = titanic_df.reset_index()
titanic_reset_df.head()

index  PassengerId  Survived  ...     Fare Cabin Embarked
0      0            1         0  ...   7.2500   NaN        S
1      1            2         1  ...  71.2833   C85        C
2      2            3         1  ...   7.9250   NaN        S
3      3            4         1  ...  53.1000  C123        S
4      4            5         0  ...   8.0500   NaN        S

[5 rows x 13 columns]

In [ ]:
#Series.reset_index()를 하게 되면 DataFrame으로 반환
print("#### before reset_index ####")
value_counts = titanic_df["Pclass"].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입 : ', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print("#### after reset_index ####")
print(new_value_counts)
print('new_value_counts 객체 변수 타입 : ', type(new_value_counts))

#### before reset_index ####
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입 :  <class 'pandas.core.series.Series'>
#### after reset_index ####
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입 :  <class 'pandas.core.frame.DataFrame'>


In [ ]:
print("#### before new_reset_index drop = False ####")
new_reset_counts = value_counts.reset_index(inplace=False,drop=False)
print("new_value_counts 객체 변수 타입 : ", type(new_reset_counts))
print(new_reset_counts)
print("#### after new_reset_index drop = True ####")
drop_reset_counts = value_counts.reset_index(drop=True)
print('drop_reset_counts 객체 변수 타입 : ', type(drop_reset_counts))
print(drop_reset_counts)

#### before new_reset_index drop = False ####
new_value_counts 객체 변수 타입 :  <class 'pandas.core.frame.DataFrame'>
   index  Pclass
0      3     491
1      1     216
2      2     184
#### after new_reset_index drop = True ####
drop_reset_counts 객체 변수 타입 :  <class 'pandas.core.series.Series'>
0    491
1    216
2    184
Name: Pclass, dtype: int64


In [10]:
#데이터 셀력션 및 필터링
#넘파이에서는 '단일값추출', '슬라이싱', '팬시', '인덱싱', '불린 인덱싱'
#판다스에서는 ix[], iloc[], loc[] 연산자

#먼저 판다스에서의 []연산자 사용
#칼럼명만 가능 
print('단일 칼럼 데이터 추출 : \n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출 : \n', titanic_df[['Survived', 'Pclass']].head(3))
#print('[]안에 숫자 index는 KeyError 오류를 발생 : \n', titanic_df[0])

#하지만 판다스에서 슬라이싱을 이용하면 가능
print("titanic_df[0:3] 슬라이싱 추출 : \n",titanic_df[0:3])

#불린 인덱싱도 가능
print("titanic_df 불린 인덱싱 추출 : \n", titanic_df[titanic_df["Pclass"] == 3].head(3))

단일 칼럼 데이터 추출 : 
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출 : 
    Survived  Pclass
0         0       3
1         1       1
2         1       3
titanic_df[0:3] 슬라이싱 추출 : 
    PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S

[3 rows x 12 columns]
titanic_df 불린 인덱싱 추출 : 
    PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
0            1         0       3  ...  7.250   NaN         S
2            3         1       3  ...  7.925   NaN         S
4            5         0       3  ...  8.050   NaN         S

[3 rows x 12 columns]


DataFrame 뒤의 [] 연산자의 입력인자의 대한 가이드
- DataFrame 바로 뒤의 [] 연산자는 Numpy의 []나 Series의 []와 다르다.
- DataFrame 바로 뒤의 [] 내 입력 값은 칼럼명(또는 칼럼리스트)을 지정해 카럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 합니다.
- DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는게 좋습니다.

In [13]:
#DataFrame의 ix[] 연산자
#곧 사라질예정이니 안쓰는것이 좋다
#print("칼럼 위치 기반 인덱싱 데이터 추출 : ", titanic_df.ix[0,2])
#print("칼럼 명 기반 인덱싱 데이터 추출 : ", titanic_df.ix[0, 'Pclass'])

#이미 사라진 것으로 보임

In [23]:
data = {"Name" : ["Chulmin", "Eunkyung", "Jinwoong","Soobeom"],
        "Year" : [2011, 2016,2015,2015],
        "Gender" : ["Male","Female","Male","Male"]}

data_df = pd.DataFrame(data, index=["one", "two", "three","four"])

#data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset = data_df.reset_index()
data_df_reset = data_df_reset.rename(columns={"index" : "old_index"})

#인덱스값에 1을 더해서 1부터 시작하느 새로운 인덱스 값 생성
data_df_reset.index = data_df_reset.index+1

data_df_reset

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

In [24]:
#DataFrame의 iloc[]
#iloc[]는 위치기반 인덱싱만 허용하기 때문에 integer 또는 integer형의 슬라이싱 팬시 리스트 값을 입력해줘야함
data_df.iloc[0,0]

'Chulmin'

In [30]:
data_df_reset.iloc[0,1]

'Chulmin'

In [48]:
#DataFrame의 loc[]
#loc[]는 명칭기반 인덱싱만 허용하기 때문에 행위치에는 DataFrame index값, 열 위치에는 칼럼명을 입력
print("인덱스가 문자열인 경우")
display(data_df)
print("data_df.loc['one','Name'] : " , data_df.loc['one','Name'])
print("\n")
print("인덱스가 정수열인 경우")
display(data_df_reset)
print("data_df_reset.loc[1,'Name'] : ", data_df_reset.loc[1,'Name'])

인덱스가 문자열인 경우


Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

data_df.loc['one','Name'] :  Chulmin


인덱스가 정수열인 경우


old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

data_df_reset.loc[1,'Name'] :  Chulmin


In [53]:
#만약 인덱스가 존재하지 않는다면
display(data_df_reset)
#print("data_df_reset.loc[0, 'Name'] : ", data_df_reset.loc[0, 'Name'])
print("data_df_reset.loc[0, 'Name'] : ","KeyError 출력")

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

data_df_reset.loc[0, 'Name'] :  KeyError 출력


In [57]:
#loc[]에서 슬라이싱을 적용하면 다른 슬라이싱과 다르게 종료값을 포함
#위치기반, 명칭기반 슬라이싱 차이점

print("위치 기반 iloc slicing \n", data_df.iloc[0:1, 0],"\n")
print("명칭 기반 loc slicing \n", data_df.loc["one":"two", "Name"])

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing 
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


In [61]:
#주의해야할 점
#loc[1:2, 'Name']에서 1:2는 DataFrame의 인덱스에 따른 명칭기반 인덱싱
display(data_df_reset)

print("data_df_reset.loc[1:2, 'Name'] : \n", data_df_reset.loc[1:2, 'Name'])

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

data_df_reset.loc[1:2, 'Name'] : 
 1     Chulmin
2    Eunkyung
Name: Name, dtype: object


iloc, loc의 문제점과 주의할 점
- 명칭기반 인덱싱과, 위치기반 인덱싱의 차이를 이해해야 합니다.
  - DataFrame의 인덱스나 칼럼명으로 데이터에 접근하는 것은 명칭 기반 인덱싱
  - 0부터 시작하는 행, 열의 위치 좌표에만 의존 하는것이 위치기반 인덱싱
- iloc[]는 위치 기반 인덱싱만 가능합니다.
  - 따라서 행과 열 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환합니다.
- loc[]는 명칭 기반 인덱싱만 가능합니다.
  - 따라서 행 위치에 DataFrame 인덱스가 오며, 열 위치에는 칼럼 명을 지정해 원하는 데이터를 반환합니다.
- 명칭 기반 인덱싱에서 슬라이싱을 **'시작점:종료점'으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환**합니다.

In [63]:
#불린 인덱싱
#[],loc[]연산자에 지원합니다. 
#로직이나 조건에 의해 계산하여 필터링을 하기 때문에 많이 사용

#승객 중 나이(Age)가 60세 이상인 데이터 추출
titanic_boolean = titanic_df[titanic_df["Age"] >= 60]
display(titanic_boolean.head())
print("titanic_boolean의 타입 : ", type(titanic_boolean))

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
33            34         0       2  ...  10.5000   NaN         S
54            55         0       1  ...  61.9792   B30         C
96            97         0       1  ...  34.6542    A5         C
116          117         0       3  ...   7.7500   NaN         Q
170          171         0       1  ...  33.5000   B19         S

[5 rows x 12 columns]

titanic_boolean의 타입 :  <class 'pandas.core.frame.DataFrame'>


In [70]:
#승객 중 나이(Age)가 60이상인 데이터의 나이(Age)와 이름(Name)을 추출
#[]연산자를 사용
print("[]연산자를 사용하는 방법")
display(titanic_df[titanic_df["Age"]>=60][["Name","Age"]].head(3))
print('\n')
print("loc[]를 사용하는 방법")
display(titanic_df.loc[titanic_df["Age"]>=60,["Name","Age"]].head(3))

[]연산자를 사용하는 방법


Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0



loc[]를 사용하는 방법


Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0

여러개의 복합 조건을 사용할 때
1. and 조건일 때는 &
2. or 조건일 때는 |
3. Not 조건일 때는 ~

In [78]:
#나이가 60세 이상이고, 선실 등급이 1등급이며, 성별이 여성인 승객 추출
#이때 개별조건은 ()로 묶고 복합 조건 연산자를 사용
#[]연사자를 사용
print("[]연산자를 사용")
display(titanic_df[(titanic_df["Age"] >= 60) &
                   (titanic_df["Pclass"] == 1) &
                   (titanic_df["Sex"] == "female")])
print('\n')
print("조건을 변수에 할당하여 사용 가능")
cond1 = titanic_df['Age'] >= 60
print("조건1 : ","cond1 = titanic_df['Age'] >= 60")
cond2 = titanic_df['Pclass'] == 1
print("조건2 : ","cond2 = titanic_df['Pclass'] == 1")
cond3 = titanic_df['Sex'] == 'female'
print("조건3 : ", "cond3 = titanic_df['Sex'] == 'female'")
display(titanic_df[cond1 & cond2 & cond3])

[]연산자를 사용


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
275          276         1       1  ...  77.9583    D7         S
366          367         1       1  ...  75.2500   D37         C
829          830         1       1  ...  80.0000   B28       NaN

[3 rows x 12 columns]



조건을 변수에 할당하여 사용 가능
조건1 :  cond1 = titanic_df['Age'] >= 60
조건2 :  cond2 = titanic_df['Pclass'] == 1
조건3 :  cond3 = titanic_df['Sex'] == 'female'


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
275          276         1       1  ...  77.9583    D7         S
366          367         1       1  ...  75.2500   D37         C
829          830         1       1  ...  80.0000   B28       NaN

[3 rows x 12 columns]

In [83]:
#정렬, Aggregation(집계)함수, GroupBy 적용

#DataFrame, Serie의 정렬 - sort_values()
#SQL의 order by 키워드와 매우 유사

print("titanic_df")
display(titanic_df.head())

titanic_sorted = titanic_df.sort_values(by=['Name'])
print("titanic_sorted")
display(titanic_sorted.head())

print("titanic_sorted by=['Pclass','Name'] ascending=False")
titanic_sorted_ = titanic_df.sort_values(by=['Pclass','Name'],ascending=False)
display(titanic_sorted_.head())

titanic_df


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

titanic_sorted


PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
845          846         0       3  ...   7.55   NaN         S
746          747         0       3  ...  20.25   NaN         S
279          280         1       3  ...  20.25   NaN         S
308          309         0       2  ...  24.00   NaN         C
874          875         1       2  ...  24.00   NaN         C

[5 rows x 12 columns]

titanic_sorted by=['Pclass','Name'] ascending=False


PassengerId  Survived  Pclass  ...  Fare Cabin  Embarked
868          869         0       3  ...   9.5   NaN         S
153          154         0       3  ...  14.5   NaN         S
282          283         0       3  ...   9.5   NaN         S
286          287         1       3  ...   9.5   NaN         S
559          560         1       3  ...  17.4   NaN         S

[5 rows x 12 columns]

In [88]:
#Aggregation(집계)함수 min(), sum(), count() etc..
#SQL의 aggregation 함수와 유사

#DataFrame에서 Aggregation

print("titanic_df.count() : ")
display(titanic_df.count())
print("DataFrame 전체에 count() 함수를 적용")

titanic_df.count() : 


PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

DataFrame 전체에 count() 함수를 적용


In [89]:
#특정칼럼에 적용 Aggregation 적용시
#특정칼럼 추출후 적용
print("titanic_df[['Age','Fare']].mean() : ")
display(titanic_df[['Age','Fare']].mean())

titanic_df[['Age','Fare']].mean() : 


Age     29.699118
Fare    32.204208
dtype: float64

In [90]:
#groupby() 적용
#SQL의 groupby와 유사하지만 다른면이 존재
#by 파라미터에 칼럼을 입력하면 대상 칼럼으로 groupby가 됨
#DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환

titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [91]:
#DataFrame에 groupby()를 호출해 반환된 결과에 aggregation 함수를 호출 하면
#groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용
titanic_groupby = titanic_df.groupby(by='Pclass').count()
display(titanic_groupby)

PassengerId  Survived  Name  Sex  ...  Ticket  Fare  Cabin  Embarked
Pclass                                    ...                               
1               216       216   216  216  ...     216   216    176       214
2               184       184   184  184  ...     184   184     16       184
3               491       491   491  491  ...     491   491     12       491

[3 rows x 11 columns]

SQL과의 다른점
- groupby 적용시 여러 개의 칼럼에 aggregation 함수를 호출하려면 대상 칼럼을 모두 Select절에 나열해야 함
  - SELECT COUNT(PassengerId), COUNT(Survived) ... FROM TITANIC_TABLE GROUP BY Pclass

In [92]:
#DataFrameGroupBy 객체에 [['PassengerId','Survived']]로 필터링
titanic_groupby = titanic_df.groupby(by='Pclass')[['PassengerId','Survived']].count()
display(titanic_groupby)

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [96]:
#SQL에서는 여러개의 aggregation 함수를 적용할 때 Select절에 나열해야함
#ex) SELECT MAX(Age), MIN(AgE) FROM TITANIC_TABLE GROUP BY Pclass

#DataFrame에서는 agg()내에 인자로 입력해서 사용
titanic_df.groupby(by='Pclass')["Age"].agg(['max','min'])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [98]:
#SQl의 groupby보다 DataFrame은 유연성이 떨어진다
#SELECT MAX(Age), MIN(SibSp), AVG(Fare) FROM TITANIC_TABLE GROUP BY Pclass
#같이 쉽게 가능함

#DataFrame의 경우
#agg()내에 딕셔너리 형태로 적용해야함
agg_format = {'Age':'max','SibSp':'min','Fare':'mean'}
titanic_df.groupby(by='Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0      0  84.154687
2       70.0      0  20.662183
3       74.0      0  13.675550

In [105]:
#결측값 처리
#결측값(Null) ==  Numpy(NaN)
#결측값은 평균, 총합등 함수 연산시 제외
#100개의 데이터중 10개가 NaN값이면 90개의 데이터에 대한 평균을 구함

#isna()함수로 결측값 확인
print("titanic_df.isna() 결측값 확인 : ")
display(titanic_df.isna().head(3))
print("True로 표시된 값이 결측값")
print("\n")
print("titanic_df의 결측값의 개수 확인 : ")
display(titanic_df.isna().sum())

titanic_df.isna() 결측값 확인 : 


PassengerId  Survived  Pclass   Name  ...  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  ...   False  False   True     False
1        False     False   False  False  ...   False  False  False     False
2        False     False   False  False  ...   False  False   True     False

[3 rows x 12 columns]

True로 표시된 값이 결측값


titanic_df의 결측값의 개수 확인 : 


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [113]:
#fillna()로 결측값 채우기
print("#### before titanic_df['Cabin'] ####")
display(titanic_df['Cabin'].head())
print("#### after titanic_df['Cabin'] ####")
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
display(titanic_df['Cabin'].head())
print("Cabin 칼럼의 NaN -> C000 확인 가능")

#### before titanic_df['Cabin'] ####


0     NaN
1     C85
2     NaN
3    C123
4     NaN
Name: Cabin, dtype: object

#### after titanic_df['Cabin'] ####


0    C000
1     C85
2    C000
3    C123
4    C000
Name: Cabin, dtype: object

Cabin 칼럼의 NaN -> C000 확인 가능


In [114]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [117]:
#apply lambda 데이터 가공
#lambda는 파이썬 함수형 프로그램 지원하기 위해 만듬
#lambda
def get_square(a):
  return a**2
print("3의 제곱은 : ",get_square(3))

#위의 식을 한줄로 간단하게 처리 가능
lambda_square = lambda x : x**2
print("3의 제곱은 : ", lambda_square(3))

3의 제곱은 :  9
3의 제곱은 :  9


In [120]:
#lambda 식을 이용할 때 여러 개의 인자값을 사용하는 경우 
#map()함수를 활용
a=[1,2,3]
squares = map(lambda x : x**2 ,a)
list(squares)

[1, 4, 9]

In [123]:
#DataFrame에 apply lambda 식 적용
titanic_df['Name_len'] = titanic_df["Name"].apply(lambda x:len(x))
titanic_df[["Name","Name_len"]].head(3)

Name  Name_len
0                            Braund, Mr. Owen Harris        23
1  Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                             Heikkinen, Miss. Laina        22

In [132]:
#lambda 식에 조건적용
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult')
titanic_df[['Age','Child_Adult']].head(10)


Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child
8  27.000000       Adult
9  14.000000       Child

In [134]:
#lamdba else if문을 지원하지 않음
#else()를 이용해야함
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

In [136]:
#조건이 많아질 경우에는  함수식을 적용하는게 유리
#나이에 따라 세분화된 분류를 수행하는 함수를 적용
def get_category(age):
  cat=''
  if age<=5 : cat = 'Baby'
  elif age <= 12 : cat = 'Child'
  elif age <= 18 : cat = 'Teenager'
  elif age <= 25 : cat = 'Student'
  elif age <= 35 : cat = 'Young Adult'
  elif age <= 60 : cat = 'Adult'
  else : cat = 'Elderly'

  return cat

#lambda 식에 위에서 생성한 get_category()함수를 반환값으로 지정

titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age','Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult